linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,180 @@
+
+If the `epoll_ctl` call fails with `ENOENT`, `completion_port_post_queue` shall call `epoll_ctl` again with `EPOLL_CTL_ADD`.
+
+On success, `completion_port_post_queue` shall return 0.
+
+If any error occurs, `completion_port_post_queue` shall fail and return a non-zero value.
+
+### completion_port_dequeue
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dequeue, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+If `completion_port` is `NULL`, `completion_port_dequeue` shall return.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_dequeue` shall return.
+
+`completion_port_dequeue` shall remove the underlying socket from the epoll by calling  `epoll_ctl` with `EPOLL_CTL_DEL`.
+
+### epoll_worker_func
+
+```c
+static int epoll_worker_func(void* parameter)
+```
+
+If `parameter` is `NULL`, `epoll_worker_func` shall do nothing.
+
+`epoll_worker_func` shall call `epoll_wait` to wait for an epoll event to become signaled.
+
+If the events value contains `EPOLLRDHUP` (hang up), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then call the `event_callback` callback with the `event_callback_ctx` and `COMPLETION_PORT_RESULT_ABANDONED`.
+
+If the events value contains `EPOLLIN` (recv), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then `epoll_worker_func` shall call `recv` and do the following:
+
+- If the recv size < 0, then:
+
+  - If `errno` is `EAGAIN` or `EWOULDBLOCK`, then unlikely errors will continue
+
+  - If `errno` is `ECONNRESET`, then `epoll_worker_func` shall call the `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_ABANDONED`
+
+- If the recv size equal 0, then `epoll_worker_func` shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
+
+- If the recv size > 0, if we have another buffer to fill then we will attempt another read, otherwise we shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
+
+If the events value contains `EPOLLOUT` (send), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- `epoll_worker_func` shall loop through the total buffers and send the data.
+
+- if send returns value is < 0 `epoll_worker_func` shall do the following:
+
+  - if `errno` is `ECONNRESET`, then `on_send_complete` shall be called with `COMPLETION_PORT_RESULT_ABANDONED`.
+
+  - if `errno` is anything else, then `on_send_complete` shall be called with `COMPLETION_PORT_RESULT_ERROR`.
```

Comment:
>is anything else [](http://example.com/codeflow?start=15&length=16)

I don't think this is right here.

I imagine a world where the socket file descriptor indicates availability of "EPOLLOUT" (writes).

And there's 100MB of data to be send.

And **some** of it might make it (send returns > 0). And some of it will not make it, because the socket's internal buffers are full.

socket internal buffers full maybe would return EWOULDBLOCK? Supposedly.